etcdserver/server.go

```diff
@@ -775,9 +775,84 @@ func (s *EtcdServer) triggerSnapshot(ep *etcdProgress) {
 	ep.snapi = ep.appliedi
 }
 
-// Stop stops the server gracefully, and shuts down the running goroutine.
-// Stop should be called after a Start(s), otherwise it will block forever.
-func (s *EtcdServer) Stop() {
+// TransferLeadership transfers the leader to the given transferee.
+func (s *EtcdServer) TransferLeadership(ctx context.Context, lead, transferee uint64) error {
+	now := time.Now()
+	interval := time.Duration(s.Cfg.ElectionTicks) * time.Millisecond
+
+	donec, stopc := make(chan struct{}), make(chan struct{})
+	go func() {
+		defer close(donec)
+
+		plog.Infof("transferring leadership from %s to %s", types.ID(lead), types.ID(transferee))
+		s.r.TransferLeadership(ctx, lead, transferee)
+		for s.Lead() != transferee {
+			select {
+			case <-stopc:
+				return
+			case <-time.After(interval):
+			}
+		}
+	}()
+
+	select {
+	case <-donec:
+	case <-ctx.Done(): // time out
+		close(stopc)
+		return ErrTimeoutLeaderTransfer
+	}
+
+	// TODO: drain all requests, or drop all messages to the old leader
+
+	plog.Infof("finished leader transfer to %s (took %v)", types.ID(transferee), time.Since(now))
+	return nil
+}
+
+// chooseLeaderTransferee chooses the member with longest active-since-time
+// as a leader transferee, since it indicates the stability of the member.
+func (s *EtcdServer) chooseLeaderTransferee() uint64 {
```

Comment:
put this in etcdserver/util.go as `longestConnected(transport rafthttp.Transporter, self types.ID, members[] *membership.Member) types.ID`?